modules/openapi-generator/src/main/resources/python/configuration.mustache

```diff
@@ -121,6 +126,17 @@ class Configuration(object):
            For RSA keys, supported values are PKCS1v15, PSS.
            For ECDSA keys, supported values are fips-186-3, deterministic-rfc6979.
         """
+        if signature_max_validity is not None and signature_max_validity.total_seconds() < 0:
+            raise Exception("The signature max validity must be a positive value")
+        self.signature_max_validity = signature_max_validity
+        """The signature max validity, expressed as a datetime.timedelta value.
+        It must be a positive value.
+        """
+        if signed_headers is not None and '(expires)' in signed_headers:
+            if self.signature_max_validity is None:
+                raise Exception("Signature max validity must be set when '(expires)' signature parameter is specified")
+        if len(signed_headers) != len(set(signed_headers)):
```

Comment:
Would it make sense to just require that a set is passed in for signed_headers?
Then we can remove this if statement.

Reply:
Here is what the spec states:

```
The list order is important, and MUST be specified in the order
the HTTP header field-value pairs are concatenated together
during Signature String Construction (Section 2.3) used during
signing and verifying.
```
It is crucial to have an ordered list *on the wire*. This is required to properly generate the signature. That being said, in principle, the Python client could provide an unordered list, as long as the implementation ensures the signature is properly generated. But I don't like this very much because on the wire, the list of signed headers will not be deterministic.

Reply:
That's good to know. Thank you for including that spec information.
Can you add code to enforce the with either of the below options?
- require that that order is passed in
- pass in a set and select items in the correct order by iterating over an ordered list of keys when generating our signed string

Reply:
Sorry, I'm not sure I understand what you are asking.

Reply:
Right now could a user could pass in a signed_headers list with items in the wrong order and our code would generate our signed string in the wrong order.
Is that true?
If so then I would like to add code which ensures that we generate the signed string using the correct order. We could do that with either of these two solutions:
- require that that the items in the input signed_headers list are in the correct order
- have our users pass in a set of signed_headers and convert it into a list which is ordered correctly inside our function

Reply:
> Right now a user could pass in a signed_headers list with items in the wrong order and our code would generate our signed string in the wrong order.
> Is that true?

No. The order is important because basically the algorithm works as follows (simplified):
1. Concatenate headers: (H1 + H2 + ... + Hn)
2. Calculate cryptographic digest of above string
3. Sign digest with private key.

The digest would be totally different if the order was different, and the client and server would not be able to agree on the signature validation.

The user may have a reason to put the list signed headers in a specific order. It's a weak argument, but certainly it helps when you are troubleshooting a problem between a client and a server. On the server side, you may not have control over the client implementation, but it helps to have easily reproducible conditions. 

> If so then I would like to add code which ensures that we generate the signed string using the correct order. We could do that with either of these two solutions:
> 
> * require that that the items in the input signed_headers list are in the correct order
> * have our users pass in a set of signed_headers and convert it into a list which is ordered correctly inside our function

Reply:
Ah, I understand now. Thank you for explaining that. So the client needs to be able to set the order and there is no correct order to use. We are all set here then. Feel free to mark this conversation as resolved.